File: NeetCode/Backtracking/N_Queens_II/n__queens__i_i.py
from typing import Optional, List
import heapq
from collections import defaultdict


class Solution:
    def totalNQueens(self, n: int) -> int:

        col = set()
        posDiag = set()
        negDiag = set()

        count = 0

        def backtrack(row):
            nonlocal count

            if row == n:
                count += 1
                return

            for c in range(n):
                if c in col or row - c in posDiag or row + c in negDiag:
                    continue

                col.add(c)
                posDiag.add(row - c)
                negDiag.add(row + c)

                backtrack(row + 1)

                col.remove(c)
                posDiag.remove(row - c)
                negDiag.remove(row + c)

        backtrack(0)

        return count

        pass


if __name__ == "__main__":
    sol = Solution()
    for i in range(10):
        print(sol.totalNQueens(i))
    print("Running Solution...")


"""  
class Solution:
    def totalNQueens(self, n: int) -> int:
        board = [["."] * n for _ in range(n)]

        columns = set()
        posDiag = set()
        negDiag = set()

        results = []

        def backtrack(row):
            if row == n:
                c = ["".join(r) for r in board]
                results.append(c)
                return
            for col in range(n):
                pDiag = row + col
                nDiag = row - col
                if col in columns or pDiag in posDiag or nDiag in negDiag:
                    continue

                columns.add(col)
                posDiag.add(pDiag)
                negDiag.add(nDiag)
                board[row][col] = "Q"

                backtrack(row + 1)

                columns.remove(col)
                posDiag.remove(pDiag)
                negDiag.remove(nDiag)
                board[row][col] = "."

        backtrack(0)
        return len(results)
"""

File: NeetCode/Backtracking/N_Queens_II/N_Queens_II.cpp
// N_Queens_II.cpp
//
// Compile locally with:
// clang++ N_Queens_II.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
private:
    int count, rows;
    unordered_set<int> columns;
    unordered_set<int> posDiag;
    unordered_set<int> negDiag;

public:
    int totalNQueens(int n)
    {
        count = 0;
        rows = n;
        backtrack(0);
        return count;
    }
    void backtrack(int row)
    {
        if (row == rows)
        {
            count++;
            return;
        }
        for (int col = 0; col < rows; col++)
        {
            if (columns.count(col) || posDiag.count(row - col) || negDiag.count(row + col))
            {
                continue;
            }

            columns.insert(col);
            posDiag.insert(row - col);
            negDiag.insert(row + col);

            backtrack(row + 1);

            columns.erase(col);
            posDiag.erase(row - col);
            negDiag.erase(row + col);
        }
    }
};

int main()
{
    cout << "Running N_Queens_II.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}
